src/sqlfluff/rules/L051.py

```diff
@@ -36,23 +43,52 @@ class Rule_L051(BaseRule):
         INNER JOIN baz;
     """
 
+    config_keywords = ["fully_qualify_join_types"]
+
     def _eval(self, context: RuleContext) -> Optional[LintResult]:
-        """INNER JOIN must be fully qualified."""
+        """Fully qualify JOINs."""
+        # Config type hints
+        self.fully_qualify_join_types: str
+
         # We are only interested in JOIN clauses.
-        if context.segment.type != "join_clause":
+        if not context.segment.is_type("join_clause"):
             return None
 
-        # We identify non-lone JOIN by looking at first child segment.
-        if context.segment.segments[0].name != "join":
-            return None
+        join_clause_keywords = [
+            segment for segment in context.segment.segments if segment.type == "keyword"
+        ]
+
+        # We identify LEFT/RIGHT/OUTER JOIN and if the next keyword is JOIN.
+        if (
+            self.fully_qualify_join_types in ["outer", "both"]
+            and join_clause_keywords[0].name in ["right", "left", "full"]
+            and join_clause_keywords[1].name == "join"
+        ):
+            # Insert OUTER after LEFT/RIGHT/OUTER
+            return LintResult(
+                context.segment.segments[0],
+                fixes=[
+                    LintFix.create_after(
+                        context.segment.segments[0],
+                        [WhitespaceSegment(), KeywordSegment("OUTER")],
```

Comment:
Should we set the case of the `OUTER` based on the case of the other keywords in case someone prefers lowercase?

Reply:
That's a great question! I'm not sure how other rules that append code when fixing linting errors behave. I'll look into that and update the code if necessary. :+1:

Reply:
Most of them try to do that where possible (see L033 as an example). And think we have the information we need here. We've had a long standing issue with one of the rules not doing that but can't find it now...

Reply:
L043.py doesn't seem to care about capitalization in its fix part.

I think that overall makes sense. We're probably better off leaving capitalization constraints and logic in the rules dedicated to that end?

Also while browsing through the code I noticed that L043.py was leveraging `raw_upper` instead of `name` which I will also adopt in this rule as it clearly answers your above question.

Reply:
> We're probably better off leaving capitalization constraints and logic in the rules dedicated to that end?

I think it depends on ho w easy it is to implement. For some of the spacing issues for example, there's a lot of logic to consider. L003 is one of our most complicated rules, so wouldn't want to duplicate that logic in other rules (it's difficult to maintain in one rule!).

However for capitalisation, when it's obvious I think we should do it in the rule fixing it. Otherwise we're just making it do extra work and another round of fixes to correct that. Also it's possible that those other rules, for `consistent` setting, it see's this rule's choice and changes the rest of the SQL to match this choice. Which would be very confusing to the end user!

So while I don't think we should implement the full capitalisation logic, I think in this case it's simple enough to copy the capitalisation of the existing keywords it's checking. Even if that's a very simplified logic (e.g. if `left` or `right` then use `outer`, otherwise use `OUTER`? Or perhaps if `join` use `outer` else use `OUTER`). WDYT?

Reply:
Your argument totally makes sense as long as we keep to a simple capitalisation logic in this rule :slightly_smiling_face: At the end of the day, you have a better overall view of architecture of this project, so I went with your suggestion of capitalising INNER/OUTER based on JOIN.

Reply:
Very nice implementation! Beautifully simple.